### ISSUE TITLE
Custom script as a deployer 

### ISSUE THREAD

**Issue by ojhughes** [GoogleContainerTools/skaffold#2277]
### Expected behavior
Ability to provide a custom script as a deployer,  using the same pattern [that allows a custom build script to be used as a builder](https://skaffold.dev/docs/how-tos/builders/#custom-build-script-run-locally). This will enable other Kubernetes deployment tools to be leveraged within a Skaffold workflow, such as [Kapp](https://get-kapp.io) or [Spinnaker pipeline triggers](https://www.spinnaker.io/guides/user/pipeline/triggers/) 

### Actual behavior
Currently, it is only possible to use `kubectl`, `helm` or `kustomize` as a deployer implementation

**Comment by NAlexPear** [comment]
This is important for us as well, since we're using [`sops`](https://github.com/mozilla/sops) for secret decryption across environments. So even though we're happy using `kubectl` to apply changes to our clusters, the inability to hook into that deploy pipeline means that we can't currently use `skaffold`

**Comment by erandagan** [comment]
More use cases:
- Using other configuration & templating solutions like [kapitan](https://github.com/deepmind/kapitan) or plain Jinja2.
- [Mixing helm and kustomize](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/chart.md) in all sorts of ways
- As @NAlexPear mentioned, handling secrets using a variety of mechanisms (Cloud KMS providers / sops / helm-secrets / ...). This is a real pain point that makes it difficult to integrate Skaffold into most of our projects even if we're already using one of the supported deployers. We're currently using the great [sealed-secrets](https://github.com/bitnami-labs/sealed-secrets) as a workaround, but would much rather use our standard KMS solution instead.

**Comment by tbarrella** [comment]
It seems like this could also reduce the need for https://github.com/GoogleContainerTools/skaffold/issues/1441

**Comment by ojhughes** [comment]
This is my current workflow
`skaffold render | kapp deploy -a $(yq r skaffold.yaml metadata.name) -y -f -`

In my case I also let [kapp](https://get-kapp.io/) take care of deletion of the deployment so would also be nice to provide an override for `skaffold delete`

**Comment by tstromberg** [comment]
Adding discuss label to see if this is actually feasible or not.

**Comment by balopat** [comment]
Adding a deploy script seems like an easy enough thing to do at first, but there needs to be careful consideration put into the design. A solution should be able to tie nicely into logging, debugging, render, port forwarding. Extracting all this behavior into an interface is not a small feature. We recognize that there would be value here - and we would love to tackle this carefully but it is at the moment not the highest on our list of things to focus on.

**Comment by horacimacias** [comment]
for logging, wouldn't it be possible to also support logging via a custom script so at least people using custom deployers can use a custom logging script?

Similar thing for debugging. 

I'm not saying it's a 5min job but at least we can allow people to craft their own scripts (which they probably already have anyway) and use the skaffold "loop".

**Comment by nkubala** [comment]
@horacimacias you're right that definitely would be possible, and in all likelihood the route we'd want to take when implementing this. as @balopat mentioned we'd need to extract all of this behavior into an interface that would be fulfilled by the custom scripts, and as he also mentioned this is definitely not a small undertaking.

we definitely hear the requests for this, and would love to see it implemented! at the moment our team's bandwidth is unfortunately already stretched pretty thin so I don't see us getting to this this quarter. designs and PRs are always welcome though 🙏

**Comment by sid-hbm** [comment]
What happened to this feature request? Was it implemented (if yes, is there any documentation about it). Thank you very much for willing to support this feature.

**Comment by aaron-prindle** [comment]
This has now been prioritized as  part of our roadmap.  We currently have no exact dates as of when to expect this but the team is now actively investigating/designing a solution here.  I believe we will have a design to submit (will be submitted in OSS to our `docs-v2/design-proposal` area) in ~3-4 weeks.  I will update the thread here w/ more information and updates

**Comment by giepa** [comment]
Has there been any progress on this?

**Comment by maulik13** [comment]
Hi, any update on this would be great. We are looking at running a few of our custom tools as part of the deployment.
